File: 41. First Missing Positive/solution.py
class Solution:
    def firstMissingPositive(self, nums: 'List[int]') -> 'int':
        # edge case
        if not nums:
            return 1
        
        # replace non-positive numbers with +infinity
        n = len(nums)
        for i in range(n):
            if nums[i] <= 0:
                nums[i] = sys.maxsize
        
        # for any number less than n, add a -ve sign to n-1th position
        for i in range(n):
            pos = abs(nums[i]) - 1
            if pos < n:
                nums[pos] = - abs(nums[pos])

        # check for a positive number at index i, and i + 1 is the solution
        for i in range(n):
            if nums[i] > 0:
                return i + 1
            
        # n + 1 is the solution if no number is found
        return n + 1

File: 41. First Missing Positive/solution.java
class Solution {
    public int firstMissingPositive(int[] nums) {
        int n = nums.length;
        if (n == 0)
            return 1;
        
        // replace neg numbers by 0
        for (int i = 0; i < n; i++) {
            if (nums[i] < 0) 
                nums[i] = 0;
        }
        
        // mark seen number by marking their index number as negative
        int index;
        for (int i = 0; i < n; i++) {
            if (Math.abs(nums[i]) <= n && nums[i] != 0) {
                index = Math.abs(nums[i]) - 1;
                if (nums[index] == 0) {
                    nums[index] = -n-1;
                } else {
                    nums[index] = -1*Math.abs(nums[index]);
                }
            }
        }
        
        // check the first index that is not marked negative
        int i;
        for (i = 0; i < n; i++) {
            if (nums[i] >= 0) {
                return i+1;
            }
        }
        
        return i+1;
    }
}
